Changed macros to work on stable (with proc-macro-hack)

danielhenrymantilla opened the pull request:
`proc-macro-hack`-based (much simpler) reimplementation of https://github.com/sfackler/rust-phf/pull/148, to address https://github.com/sfackler/rust-phf/issues/106

sfackler commented:
@dtolnay is the missing documentation warning expected?

dtolnay commented:
Yes, you have to write the documentation. ;)

```rust
/// ...
#[cfg(feature = "macros")]
#[proc_macro_hack]
pub use phf_macros::phf_map;

...
```

Speedy37 commented:
What is blocking you from merging this and stabilizing it?

abonander commented on phf/Cargo.toml:
```diff
@@ -1,7 +1,7 @@
 name = "phf"
 authors = ["Steven Fackler <sfackler@gmail.com>"]
-version = "0.7.24"
+version = "0.7.25"
```
I feel like it's not really in the PR author's purview to edit the crate version number. I think the next release is going to be 0.8.0 anyway.

danielhenrymantilla commented:
Reverted the version bump, rebased on master, and added some documentation on the macros.

Also added a field in `Cargo.toml` to ensure that the macros documentation appears on https://docs.rs.

Remaining question: should the `"macros"` feature by enabled by default? I haven't done it, but except maybe for compatibility with older `rustc` compilers, I can't see a reason for them no to be enabled by default, given that they are stable.

abonander commented:
We've technically bumped the minimum to 1.31 now but that's mostly to use `const fn` on stable with `unicase`. Upgrading `siphasher` apparently needed a new minimum version but I'm not sure what it was. So `macros` on by default should be fine; I'd say we could remove the feature flag entirely but if people aren't going to use it then they might as well be able to turn it off to save a little bit of compile time.

abonander commented:
I can make that change myself, thanks for your contribution!

danielhenrymantilla commented:
I'm glad I could help. Now updating the `README.md` would be a good thing to do :)

abonander commented:
Yep.
